lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py

```diff
@@ -0,0 +1,472 @@
+            self.module.fail_json(
+                msg='Failure validating import spec: %s' % joined_errors
+            )
+
+        for warning in getattr(self.import_spec, 'warning', []):
+            self.module.warn(to_native(warning.msg))
+
+        try:
+            self.lease = resource_pool.ImportVApp(
+                self.import_spec.importSpec,
+                datacenter.vmFolder
+            )
+        except vmodl.fault.SystemError as e:
+            self.module.fail_json(
+                msg='Failed to start import: %s' % to_native(e.msg)
+            )
+
+        while self.lease.state != vim.HttpNfcLease.State.ready:
+            time.sleep(0.1)
+
+        self.entity = self.lease.info.entity
+
+        return self.lease, self.import_spec
+
+    def upload(self):
+        ovf_dir = os.path.dirname(self.params['ovf'])
+
+        lease, import_spec = self.get_lease()
+
+        for file_item in import_spec.fileItem:
+            vmdk_post_url = None
+            for device_url in lease.info.deviceUrl:
+                if file_item.deviceId == device_url.importKey:
+                    vmdk_post_url = device_url.url.replace('*', self.params['hostname'])
+                    break
+
+            if not vmdk_post_url:
+                lease.HttpNfcLeaseAbort(
+                    vmodl.fault.SystemError(reason='Failed to find deviceUrl for file %s' % file_item.path)
+                )
+                self.module.fail_json(
+                    msg='Failed to find deviceUrl for file %s' % file_item.path
+                )
+
+            if self.tar:
+                vmdk = self.tar
+                try:
+                    vmdk_tarinfo = self.tar.getmember(file_item.path)
+                except KeyError:
+                    lease.HttpNfcLeaseAbort(
+                        vmodl.fault.SystemError(reason='Failed to find VMDK file %s in OVA' % file_item.path)
+                    )
+                    self.module.fail_json(
+                        msg='Failed to find VMDK file %s in OVA' % file_item.path
+                    )
+            else:
+                vmdk = os.path.join(ovf_dir, file_item.path)
+                try:
+                    path_exists(vmdk)
+                except ValueError:
+                    lease.HttpNfcLeaseAbort(
+                        vmodl.fault.SystemError(reason='Failed to find VMDK file at %s' % vmdk)
+                    )
+                    self.module.fail_json(
+                        msg='Failed to find VMDK file at %s' % vmdk
+                    )
+                vmdk_tarinfo = None
+
+            uploader = VMDKUploader(
+                vmdk,
+                vmdk_post_url,
+                self.params['validate_certs'],
+                tarinfo=vmdk_tarinfo
+            )
+            uploader.start()
+            while uploader.is_alive():
+                time.sleep(0.1)
+                lease.HttpNfcLeaseProgress(int(100.0 * uploader.bytes_read / uploader.size))
```

Comment:
There might be a bug calculating progress here.  I may need to create all the uploaders in the for loop but not start them, calculate total size, and then process the uploaders in their own loop, tracking total uploaded percent.

I'm guessing, based on a report, that this will go to 100, then on the next file attempt to reset to 0, which the API doesn't allow.